src/librustc_mir/build/cfg.rs

```diff
@@ -86,12 +86,17 @@ impl<'tcx> CFG<'tcx> {
                      scope: ScopeId,
                      span: Span,
                      kind: TerminatorKind<'tcx>) {
-        debug_assert!(self.block_data(block).terminator.is_none(),
+        debug_assert!(!self.terminated(block),
                       "terminate: block {:?} already has a terminator set", block);
         self.block_data_mut(block).terminator = Some(Terminator {
             span: span,
             scope: scope,
             kind: kind,
         });
     }
+
+    /// Returns whether or not the given block has been terminated or not
+    pub fn terminated(&self, block: BasicBlock) -> bool {
```

Comment:
Essentially one shouldn’t ever have to check whether block is terminated in the builder, with a notable exception of the `terminate` itself.